loading-bay/camera_manager.py

```diff
@@ -0,0 +1,162 @@
+        Returns:
+            A cv2 VideoSource
+        """
+        camera = self.cs.startAutomaticCapture(name=config["name"], path=config["path"])
+        camera.setConfigJson(json.dumps(config["config"]))
+        return camera
+
+    def get_frame(self, camera: int = 0) -> tuple:
+        """Gets a frame from the specified camera.
+
+        Args:
+            camera: Which camera to get the frame from. Default is 0.
+        Returns:
+            Frame_time, or 0 on error.
+            A numpy array of the frame, dtype=np.uint8, BGR.
+        """
+        frame_time, self.frame = self.sinks[camera].grabFrameNoTimeout(image=self.frame)
+        return frame_time, self.frame
+
+    def send_frame(self, frame: np.ndarray) -> None:
+        """Sends a frame to the driver display.
+
+        Args:
+            frame: A numpy array image. (Should always be the same size)
+        """
+        self.source.putFrame(frame)
+
+
+class DummyImageManager:
+    def __init__(self, image: np.ndarray) -> None:
+        """Initialises a Dummy Image Manager
+
+        Args:
+            image: A BGR numpy image array
+        """
+        self.image = image
+
+    def change_image(self, new_image: np.ndarray) -> None:
+        """Changes self.image.
+
+        Args:
+            new_image: The new image to switch to.
+        """
+        self.image = new_image
+
+    def get_frame(self, camera: int = 0) -> tuple:
+        """Returns self.image.
+
+        Args:
+            camera: Not needed, just here to ensure similarity with CameraManager.
+        Returns:
+            0.1: Simulates the frame_time
+            self.image, a BGR numpy array.
+        """
+        return 0.1, self.image
+
+    def send_frame(self, frame: np.ndarray):
+        ...
+
+
+class DummyVideoManager:
+    def __init__(self, video: cv2.VideoCapture):
+        """Initialises a Dummy Video Manager.
+
+        Args:
+            video: An opencv video, as received by cv2.VideoCapture
+        """
+        self.video = video
+
+    def get_frame(self, camera: int = 0) -> tuple:
+        """Returns the next frame of self.video.
+
+        Args:
+            camera: Not needed, just here to ensure similarity with CameraManager.
+        Returns:
+            Whether or not it was successful. False means error.
+            The next frame of self.video.
+        """
+        result = self.video.read()
+        if result[0]:
+            return result
+        else:  # If we reach the end of the video, go back to the beginning.
+            self.video.set(cv2.CAP_PROP_POS_FRAMES, 0)
+            return self.get_frame()
+
+    def send_frame(self, frame: np.ndarray) -> None:
+        ...
+
+
+class DummyCameraManager:
+    def __init__(self, camera: int = 0) -> None:
+        """Initialises a Dummy Camera Manager. Designed to run on a non-pi computer.
+        Initialises it with the first detected system camera, for example a webcam.
```

Comment:
If it's a dummy, why does it use a real camera? Call it WebcamCameraManager, or something, but not "dummy". It's not a mock, either, because it uses live, i.e. uncontrolled, data.